db/migrate/20210414203512_convert_primary_language_to_iso_code.rb

```diff
@@ -0,0 +1,138 @@
+      unknown_languages << el if matched_lang.nil?
+      automatic_translations[el.to_sym] = matched_lang&.to_s
+    end
+
+    ActiveRecord::Base.transaction do
+      automatic_translations.each do |key, value|
+        key = key.to_s
+        Patient.where(purged: false).where(primary_language: key).update_all(primary_language: value)
+        insert_history_items(key, value, true) if unknown_languages.include?(key)
+        Patient.where(purged: false).where(secondary_language: key).update_all(secondary_language: value)
+        insert_history_items(key, value, false) if unknown_languages.include?(key)
+      end
+
+      custom_translations.each do |key, value|
+        key = key.to_s
+        case value
+        when 'bilingual english/spanish'
+          insert_history_items(key, 'eng', true) if TRANSLATION_COMMENTS.include?(key)
+          Patient.where(purged: false).where(primary_language: key).update_all({ primary_language: 'eng', secondary_language: 'spa' })
+        when 'bilingual spanish/english'
+          insert_history_items(key, 'spa', true) if TRANSLATION_COMMENTS.include?(key)
+          Patient.where(purged: false).where(primary_language: key).update_all({ primary_language: 'spa', secondary_language: 'eng' })
+        else
+          insert_history_items(key, value, true) if TRANSLATION_COMMENTS.include?(key)
+          Patient.where(purged: false).where(primary_language: key).update_all(primary_language: value)
+          insert_history_items(key, value, false) if TRANSLATION_COMMENTS.include?(key)
+          Patient.where(purged: false).where(secondary_language: key).update_all(secondary_language: value)
+        end
+      end
+    end
+  end
+
+  def down
+    execute <<-SQL.squish
+        UPDATE patients SET primary_language = legacy_primary_language, secondary_language = legacy_secondary_language
+    SQL
+
+    remove_column :patients, :legacy_primary_language
+    remove_column :patients, :legacy_secondary_language
+
+    # Destroy all System Record Edits that contain the text `language was listed as`
+    History.where('history_type = ? AND comment like ?', 'System Record Edit', '%language was listed as%').destroy_all
+  end
+
+  def insert_history_items(key, value, is_primary)
+    note = "#{is_primary ? 'Primary' : 'Secondary'} language was listed as '#{key}' \
+    which could not be matched to the standard language list. This monitoree’s #{is_primary ? 'primary' : 'secondary'} \
+    language has been updated to '#{value.nil? ? 'blank' : Languages.all_languages[value.to_sym][:display]}'. You \
+    may update this value."
+    execute <<-SQL.squish
+      INSERT INTO histories (patient_id, created_by, comment, history_type, created_at, updated_at)
+      SELECT id, 'Sara Alert System', "#{note}", 'System Record Edit', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP
+      FROM patients
+      WHERE purged = false AND #{is_primary ? 'primary_language' : 'secondary_language'} = "#{key}"
```

Comment:
There is a logic error with this line when the subject of the history item has a previously unknown language. Since the language was already updated to `nil` the where needs to look for patients with a legacy_language of key.

Reply:
Ahh good catch! I'll move the translations below the insert_history_items.